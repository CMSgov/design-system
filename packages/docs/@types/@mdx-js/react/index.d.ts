declare module '@mdx-js/react' {
  const MDXProvider: React.ComponentType;
  export { MDXProvider };
}
