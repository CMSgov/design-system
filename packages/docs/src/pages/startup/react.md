---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
