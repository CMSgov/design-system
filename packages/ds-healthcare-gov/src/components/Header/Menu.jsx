/* eslint-disable jsx-a11y/no-redundant-roles */
import MenuLinks from './MenuLinks';
import PropTypes from 'prop-types';
import React from 'react';
import classnames from 'classnames';

const Menu = function (props) {
  const classes = classnames('hc-c-menu', {
    'hc-c-menu--open': props.open,
  });

  return (
    <nav id="hc-c-menu" className={classes} role="navigation">
      <div className="ds-l-container">
        <div className="ds-l-row">
          <div className="hc-c-menu__content ds-u-padding--1 ds-u-font-weight--bold">
            {props.submenuTop}
            {props.beforeLinks}
            {props.links && (
              <div className="ds-l-col ds-l-col--auto ds-u-padding-x--0">
                <MenuLinks links={props.links} />
              </div>
            )}
            {props.submenuBottom}
          </div>
        </div>
      </div>
    </nav>
  );
};

Menu.propTypes = {
  links: PropTypes.arrayOf(
    PropTypes.shape({
      identifier: PropTypes.string,
      href: PropTypes.string.isRequired,
      label: PropTypes.node.isRequired,
      onClick: PropTypes.func,
    })
  ),
  submenuTop: PropTypes.node,
  submenuBottom: PropTypes.node,
  /**
   * Nodes to be rendered before the links column
   */
  beforeLinks: PropTypes.node,
  /**
   * When the menu is collapsed, passing in "open" will
   * expand it and make it visible.
   */
  open: PropTypes.bool,
};

export default Menu;
