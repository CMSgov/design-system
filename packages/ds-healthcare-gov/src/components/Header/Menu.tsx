/* eslint-disable jsx-a11y/no-redundant-roles */
import MenuLinks from './MenuLinks';
import React from 'react';
import classnames from 'classnames';
import { DefaultLink } from './defaultMenuLinks';
import { Link } from './Header';

interface MenuProps {
  links?: Array<Link | DefaultLink>;
  submenuTop?: React.ReactNode;
  submenuBottom?: React.ReactNode;
  /**
   * Nodes to be rendered before the links column
   */
  beforeLinks?: React.ReactNode;
  /**
   * When the menu is collapsed, passing in "open" will
   * expand it and make it visible.
   */
  open?: boolean;
}

const Menu = (props: MenuProps) => {
  const classes = classnames('hc-c-menu', {
    'hc-c-menu--open': props.open,
  });

  return (
    <nav id="hc-c-menu" className={classes} role="navigation">
      <div className="ds-l-container">
        <div className="ds-l-row">
          <div className="hc-c-menu__content ds-u-padding--1 ds-u-font-weight--bold">
            {props.submenuTop}
            {props.beforeLinks}
            {props.links && (
              <div className="ds-l-col ds-l-col--auto ds-u-padding-x--0">
                <MenuLinks links={props.links} />
              </div>
            )}
            {props.submenuBottom}
          </div>
        </div>
      </div>
    </nav>
  );
};

export default Menu;
