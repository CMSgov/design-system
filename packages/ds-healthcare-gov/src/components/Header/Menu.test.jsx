import Menu from './Menu';
import React from 'react';
import { render } from '@testing-library/react';

describe('Menu', function () {
  const links = [
    { href: '#foo', label: 'Foo' },
    { href: '#bar', label: 'Bar' },
  ];

  it('has an id for ARIA attributes', () => {
    // Note: the ID attribute of this component is important and used as an
    // ARIA attribute in ActionMenu. Be mindful of this if you change it.
    const { container } = render(<Menu links={links} />);

    expect(container.firstChild.id).toBe('hc-c-menu');
  });

  it('renders MenuList', () => {
    const { container } = render(<Menu links={links} />);
    expect(container).toMatchSnapshot();
  });

  it('removes hidden state', () => {
    const { container } = render(<Menu open links={links} />);
    expect(container).toMatchSnapshot();
  });
});
