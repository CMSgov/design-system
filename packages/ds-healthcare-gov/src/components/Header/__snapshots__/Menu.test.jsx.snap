// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu removes hidden state 1`] = `
<nav
  className="hc-c-menu hc-c-menu--open"
  id="hc-c-menu"
  role="navigation"
>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row"
    >
      <div
        className="hc-c-menu__content ds-u-padding--1 ds-u-font-weight--bold"
      >
        <div
          className="ds-l-col ds-l-col--auto ds-u-padding-x--0"
        >
          <MenuLinks
            links={
              Array [
                Object {
                  "href": "#foo",
                  "label": "Foo",
                },
                Object {
                  "href": "#bar",
                  "label": "Bar",
                },
              ]
            }
          />
        </div>
      </div>
    </div>
  </div>
</nav>
`;

exports[`Menu renders MenuList 1`] = `
<nav
  className="hc-c-menu"
  id="hc-c-menu"
  role="navigation"
>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row"
    >
      <div
        className="hc-c-menu__content ds-u-padding--1 ds-u-font-weight--bold"
      >
        <div
          className="ds-l-col ds-l-col--auto ds-u-padding-x--0"
        >
          <MenuLinks
            links={
              Array [
                Object {
                  "href": "#foo",
                  "label": "Foo",
                },
                Object {
                  "href": "#bar",
                  "label": "Bar",
                },
              ]
            }
          />
        </div>
      </div>
    </div>
  </div>
</nav>
`;
